### ISSUE TITLE
Expose limits imposed by cgroups via port library APIs

### ISSUE THREAD

**Issue by ashu-mehra** [eclipse/omr#1281]
Runtimes like JVM rely on resource limits like memory and cpus to adjust their internal parameters which includes heap size, GC helper threads, JIT helper threads etc.
Cgroups in Linux provide a mechanism to impose limits on such resources. Therefore, when running as part of a cgroup, the runtime should adjust its internal parameters based on the resource constraints imposed by the cgroup.
Currently the information about resource constraints imposed by cgroup is not provided by the port library in any way. If a runtime wants to take into consideration the limits imposed by the cgroup, it would have to figure this out on its own.
It would be useful for the runtimes if port library can provide APIs that would enable them to query the resource constraints imposed by the cgroup.
The proposal is to add mechanism for following tasks:
1. Add capability to read `/proc/<pid>/cgroup` file.
Using this file, the process can figure out the cgroup it belongs to for a given subsystem.
2. Expose APIs for providing memory limits and CPU limits imposed by the cgroup. For start only these two limits would be exposed. More can be added later as and when required.
3. Now, the above mechanism would work when running on bare metal or a VM, but would break when running in a docker container. This is because in a docker container the `/proc/<pid>/cgroup` gives the cgroup of the process on the host, and not the one in the container. AFAIK when running in container, the cgroup would always be the root cgroup. Therefore, we also need to detect whether the process is running in a docker container or not, and use correct cgroup to get the resource limits.

As an example, if I log in to a docker container, and cat `/proc/1/cgroup`, I get following contents:
```
$ docker exec -it dfee854963d4575d263785ddee0f3c398355b354e6608c5777a3a2e40fef3612 /bin/bash
root@dfee854963d4:/# cat /proc/1/cgroup
11:pids:/docker/dfee854963d4575d263785ddee0f3c398355b354e6608c5777a3a2e40fef3612
10:perf_event:/docker/dfee854963d4575d263785ddee0f3c398355b354e6608c5777a3a2e40fef3612
9:freezer:/docker/dfee854963d4575d263785ddee0f3c398355b354e6608c5777a3a2e40fef3612
8:cpuset:/docker/dfee854963d4575d263785ddee0f3c398355b354e6608c5777a3a2e40fef3612
7:blkio:/docker/dfee854963d4575d263785ddee0f3c398355b354e6608c5777a3a2e40fef3612
6:memory:/docker/dfee854963d4575d263785ddee0f3c398355b354e6608c5777a3a2e40fef3612
5:hugetlb:/docker/dfee854963d4575d263785ddee0f3c398355b354e6608c5777a3a2e40fef3612
4:devices:/docker/dfee854963d4575d263785ddee0f3c398355b354e6608c5777a3a2e40fef3612
3:net_prio,net_cls:/docker/dfee854963d4575d263785ddee0f3c398355b354e6608c5777a3a2e40fef3612
2:cpuacct,cpu:/docker/dfee854963d4575d263785ddee0f3c398355b354e6608c5777a3a2e40fef3612
1:name=systemd:/docker/dfee854963d4575d263785ddee0f3c398355b354e6608c5777a3a2e40fef3612
```

which shows the process is part of `/docker/dfee854963d4575d263785ddee0f3c398355b354e6608c5777a3a2e40fef3612` cgroup in memory subsystem. However, if I go to `/sys/fs/cgroup/memory` I don't see any such cgroup, but the same cgroup is present in the host. Probably the `/sys/fs/cgroup/memory/docker/<id>` cgroup on the host is mapped as the root cgroup inside the container.
In short, `/proc/<pid>/cgroup` cannot be used for getting the cgroup when running in a docker container.

**Comment by ashu-mehra** [comment]
fyi @charliegracie @DanHeidinga

**Comment by DanHeidinga** [comment]
Thanks @ashu-mehra for starting on this - it's an important feature and definitely needed in OMR!

The API for this should make it easy to make existing code cgroup-aware without having to explicitly code a cgroup vs non-cgroup path in all code requesting i.e. sys max memory.

A very rough sketch of what I'm thinking would be:

```
/* Determine if running in a cgroup namespace */
BOOLEAN omr_sysinfo_cgroup_in_cgroup()

/* Indicate to port library that limits should be taken from the cgroup, not from the machine.
 * @return previous value of this setting
 */
BOOLEAN omr_sysinfo_cgroup_enable_cgroup_limits()

/* Should the port library respect cgroups in its normal API calls */
BOOLEAN omr_sysinfo_cgroup_limits_enabled()

/* Special accessors to get the cgroup-related limit for YYYY.
 * ie: long omr_sysinfo_cgroup_get_max_memory
 */
XXXX omr_sysinfo_cgroup_get_max_YYYYY()
```

And existing apis would be updated to be cgroup aware:
```
XXXX omr_sysinfo_get_max_memory() {
   if (omr_sysinfo_cgroup_limits_enabled()) {
      return omr_sysinfo_get_max_memory();
   } else {
      ....
      /* existing code path */
      .....
   }
```

This makes it easy for all consuming code to be aware of cgroup limits without having to be modified.  Does this seem like a reasonable (though very rough) api sketch?

**Comment by ashu-mehra** [comment]
@DanHeidinga Thanks for jotting down the structure. I have exactly similar usage pattern in mind and I agree existing APIs should be updated to use cgroups when available.

Couple of points:
1. Do we need to make `omr_sysinfo_cgroup_limits_enabled()` an API, or just an internal helper function?
2. Same for `omr_sysinfo_cgroup_in_cgroup()`. It is checking if we are cgroup or not, which would be needed internally by `omr_sysinfo_cgroup_enable_cgroup_limits()`. So would it be better to just have it as helper function?

**Comment by DanHeidinga** [comment]
> Do we need to make `omr_sysinfo_cgroup_limits_enabled()` an API, or just an internal helper function?

My gut says to make it an API so that reporting / monitoring tools such as a javacore file can report which mode the port library has been put in.

> Same for `omr_sysinfo_cgroup_in_cgroup().`

For the same kinds of reasons, we want this to be query-able by the runtime using the port library.

And to be clear, I'm not sold on the method names proposed in the rough sketch.  If you think of something better, go with it - these were off the cuff examples.

**Comment by ashu-mehra** [comment]
As there are quite a few things to be done to fully support cgroups, I have tried to break it down into smaller items and created a task list in the first comment.
I will try to create pull requests targeting these tasks in the list.

If I have missed anything, feel free to add tasks in the list in the first comment.

**Comment by ashu-mehra** [comment]
@DanHeidinga @charliegracie @dinogun

For the task of determining if we are running inside a container or not, I have following thoughts:

Since there are multiple container types, not just Docker (Rkt is another going around), having a port library API omrsysinfo_get_container_type() would be relevant. 
Detecting container type is not straight forward and not well documented. We would have to use all sorts of mechanisms to figure that out. 
We can propose to set an env variable "CONTAINER_TYPE" by the user which would indicate the container in which the runtime is running.
For Docker, mechanism described in this [stackeoverflow thread](https://stackoverflow.com/questions/20010199/determining-if-a-process-runs-inside-lxc-docker) can be used:
`The most reliable way is to check /proc/1/cgroup. It will tell you the control groups of the init process, and when you are not in a container, that will be / for all hierarchies. When you are inside a container, you will see the name of the anchor point; which, with LXC/Docker containers, will be something like /lxc/<containerid> or /docker/<containerid> respectively.`

Based on above thoughts I have following code structure. Let me know if this makes sense.
```
enum ContainerType {
	DOCKER,
	RKT,
        ...,
	UNSUPPORTED,
	NONE
}
ContainerType omrsysinfo_get_container_type()

omrsysinfo_get_container_type() {
	type = getenv("CONTAINER_TYPE");
	if (NULL != type) {
		if (type == "docker" )
			return DOCKER;
		else if (type == "rkt")
			return RKT;
		else 
			return UNSUPPORTED;
	} else {
		if (isDockerContainer()) {
			return DOCKER;
		} else if (isRktContainer()) {
			return RKT;
		} else {
			return NONE;
		}
	}
}

isDockerContainer() {
	cgroup = getCgroupForPID(1)
	if (cgroup == "/") {
		return FALSE;
	} else if (cgroup.contains("docker")) {
		return TRUE;
	}
}

isRktContainer() {
      ...
}
```

**Comment by dinogun** [comment]
+1. 
This is as per the discussion that we had earlier today and makes sense to me. This is also how we have designed the isVirtual API as well.

**Comment by EricYangIBM** [comment]
Reusing this issue as a tracker for cgroup v2, prompted by https://github.com/eclipse-openj9/openj9/issues/6811. First steps will be to fix https://github.com/eclipse-openj9/openj9/issues/14190 and running in container detection by updating `omrsysinfo_cgroup_is_system_available` (and `isRunningInContainer` and `readCgroupFile` helpers).

**Comment by babsingh** [comment]
### Enabling cgroup v2 on Linux

```
 sudo grubby --update-kernel=ALL --args="systemd.unified_cgroup_hierarchy=1"
```

If grubby is not available, edit the `GRUB_CMDLINE_LINUX` line in `/etc/default/grub`:
```
GRUB_CMDLINE_LINUX="systemd.unified_cgroup_hierarchy=1"
```

The host needs to be rebooted. Once cgroup v2 is enabled, all Docker containers created on the host will utilize cgroup v2.

**Comment by babsingh** [comment]
### Identifying cgroup v2

#### Method 1

If `/sys/fs/cgroup/cgroup.controllers` is available on the host, we can assume the host is using cgroup v2. 

**Ref:** https://docs.docker.com/config/containers/runmetrics/#enumerate-cgroups.

```
# cat /sys/fs/cgroup/cgroup.controllers

cpu io memory pids rdma // List of available controllers
```

#### Method 2

Derive if a host is using cgroup v2 from `/proc/cgroups`.
For cgroup v2, all controllers should have `0` for `hierarchy`.

It should be noted that `hierarchy` can also be `0` if 
- the controller is not mounted on a cgroups v1 hierarchy; and
- the controller is disabled.

**Ref:** Search for `/proc/cgroups` in https://man7.org/linux/man-pages/man7/cgroups.7.html.

```
cat /proc/cgroups

#subsys_name	hierarchy	num_cgroups	enabled
cpuset	0	159	1
cpu	0	159	1
cpuacct	0	159	1
blkio	0	159	1
memory	0	159	1
devices	0	159	1
freezer	0	159	1
net_cls	0	159	1
perf_event	0	159	1
net_prio	0	159	1
hugetlb	0	159	1
pids	0	159	1
rdma	0	159	1
````

**Comment by babsingh** [comment]
### Differences in /proc/[pid]/cgroup between cgroup v1 and v2

For cgroup v1,
- there are multiple lines in `/proc/[pid]/cgroup`;
- format of each line: `hierarchy-ID:controller-list:cgroup-path`; and
- sample output is shown in https://github.com/eclipse/omr/issues/1281#issue-237512523.

For cgroup v2,
- there is a single line in `/proc/[pid]/cgroup`;
- format: `0::$PATH`; and
- sample output:
```
# cat /proc/2381/cgroup

0::/user.slice/user-119.slice/user@119.service/dbus.service
```

**Ref:** Search for `/proc/[pid]/cgroup` in https://man7.org/linux/man-pages/man7/cgroups.7.html.
```
For each cgroup hierarchy of which the process is a
member, there is one entry containing three colon-
separated fields:
             hierarchy-ID:controller-list:cgroup-path

The colon-separated fields are, from left to right:
1. For cgroups version 1 hierarchies, this field contains
 a unique hierarchy ID number that can be matched to a
 hierarchy ID in /proc/cgroups.  For the cgroups version
 2 hierarchy, this field contains the value 0.

2. For cgroups version 1 hierarchies, this field contains
 a comma-separated list of the controllers bound to the
 hierarchy.  For the cgroups version 2 hierarchy, this
 field is empty.

3. This field contains the pathname of the control group
 in the hierarchy to which the process belongs.  This
 pathname is relative to the mount point of the
 hierarchy.
```

**Comment by EricYangIBM** [comment]
I believe the reason we cannot detect that we are running in a container for cgroup2 is in [isRunningInContainer](https://github.com/eclipse/omr/blob/cd8a2e017030701fc5b39ea162acdec128c6aecb/port/unix/omrsysinfo.c#L5879). `isCgroupV1Available` fails for cgroup2 as `/sys/fs/cgroup` is not of `tmpfs` type:
```
root@ee74c8987adf:/# df -T /sys/fs/cgroup/
Filesystem     Type    1K-blocks  Used Available Use% Mounted on
cgroup         cgroup2         0     0         0    - /sys/fs/cgroup
```
We can add a check for cgroup2 using the above comments.

After the `isCgroupV1Available` check:
```
/* Read PID 1's cgroup file /proc/1/cgroup and check cgroup name for each subsystem.
 * If cgroup name for each subsystem points to the root cgroup "/",
 * then the process is not running in a container.
 * For any other cgroup name, assume we are in a container.
 */
```
This doesn't seem to apply to cgroup2:
```
root@ee74c8987adf:/# cat /proc/1/cgroup
0::/
```
So we will need a different algorithm for determining if the process is running in a container for cgroup2.

**Comment by babsingh** [comment]
> different algorithm for determining if the process is running in a container for cgroup2

https://github.com/GoogleContainerTools/kaniko/issues/1592 covers container detection with cgroup v2. They try to look for `/.dockerenv` to determine if the process is running in a container. [Not a robust solution](https://github.com/GoogleContainerTools/kaniko/issues/1592#issuecomment-861254141) since `/.dockerenv` may not be always available. So, there are fallback paths to identify other container specific characteristics such as `/run/.containerenv`, `/run/systemd/container`, `env("container")`, etc.

**Comment by babsingh** [comment]
re https://github.com/eclipse/omr/pull/6422#issuecomment-1071222912:
```
1CICGRPINFO    JVM support for cgroups enabled : FALSE
```

The above line is `FALSE` since `omrsysinfo_cgroup_is_system_available` has not been updated yet.

In `omrsysinfo_cgroup_is_system_available`, we generate a list of `OMRCgroupEntry` -> `PPG_cgroupEntryList`. For cgroup v2, we will need to populate this list by reading `/proc/cgroups` and `/proc/[pid]/cgroup` (pid=self). `hierarchyId` and `subsystem` are reported in `/proc/cgroups`, and `cgroup` name can be derived from `/proc/self/cgroup` and should be the same for all entries. Will it be of value to store info about `num_cgroups` and `enabled` from `/proc/cgroups`  into `OMRCgroupEntry`?

The above approach for cgroup v2 should also work with cgroup v1.

```
typedef struct OMRCgroupEntry {
	int32_t hierarchyId; /**< cgroup hierarch ID*/
	char *subsystem; /**< name of the subsystem*/
	char *cgroup; /**< name of the cgroup*/
	uint64_t flag; /**< a bit-wise flag of type OMR_CGROUP_SUBSYSTEM_* representing the cgroup*/
	struct OMRCgroupEntry *next; /**< pointer to next OMRCgroupEntry*/
} OMRCgroupEntry;
```

**Comment by EricYangIBM** [comment]
https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#deprecated-v1-core-features
> /proc/cgroups is meaningless for v2. Use “cgroup.controllers” file at the root instead.

In v1 /proc/[pid]/cgroup told us if a process had certain controllers enabled. We could then get the e.g. memory stats with `/sys/fs/cgroup/memory/cgroupName/memory.limit_in_bytes`. (see `getAbsolutePathOfCgroupSubsystemFile`)

 In v2 we can't do that so to tell if a controller is enabled for [pid] we would have to read /proc/[pid]/cgroup to get the cgroupName, then read `/sys/fs/cgroup/cgroupName/cgroup.controllers` to see if memory is available for cgroupName (the `memory.` files are available if yes). We can then get memory.max from `/sys/fs/cgroup/cgroupName/memory.max`. So we don't need `char *subsystem;` in `OMRCgroupEntry` for v2.

**Comment by babsingh** [comment]
@EricYangIBM you are correct; we should use `cgroup.controllers` for cgroup v2. so, the new algorithm will look like
```
cgroupPath <- derive from /proc/[pid]/cgroup or /proc/self/cgroup; format = 0::$cgroup_path
controller_list <- $mount_point/$cgroup_path/cgroup.controllers
foreach controller in controller_list:
   OMRCgroupEntry entry <- {hierarchyId = 0; subsystem = controller; cgroupPath = cgroupPath; flag <- ...}
   Append entry to PPG_cgroupEntryList
```

Based upon the doc, `cgroupPath` will be a more accurate naming instead of `cgroupName`. it is relative to the mount point, which we assume to be always `/sys/fs/cgroup`. I don't think this assumption is always true. can you verify this? if not true, we will need to find a way to derive the mount point.

**Comment by babsingh** [comment]
### retrieveLinuxCgroupMemoryStats

In cgroup v1, the following details are acquired (ref: https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt):
```
- memoryLimit <- $mount_point/$subsytem_name[=memory]/$cgroup_path/memory.limit_in_bytes (file only contains the numerical value)
- memoryUsage <- .../memory.usage_in_bytes (file only contains the numerical value)
- memoryAndSwapLimit <- .../memory.memsw.limit_in_bytes (file only contains the numerical value)
- memoryAndSwapUsage <- .../memory.memsw.usage_in_bytes (file only contains the numerical value)
- cached <- From .../memory.stat, greps for "cached %VALUE%" (# of bytes of page cache memory)
```

**... TRANSLATION from cgroup v1 to v2 ...**

In cgroup v2, the identical transformations will be as follows (ref: https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#memory):
```
- memoryLimit <- $mount_point/$cgroup_path/memory.max (file only contains the numerical value)
- memoryUsage <- .../memory.current (file only contains the numerical value)
- memoryAndSwapLimit <- .../memory.swap.max (file only contains the numerical value)
- memoryAndSwapUsage <- .../memory.swap.current (file only contains the numerical value)
- cached <- .../memory.stat, grep for "file %VALUE%" [unsure about this one; reconfirm!!!]
```

**Comment by babsingh** [comment]
### omrsysinfo_get_number_CPUs_by_type

https://github.com/eclipse/omr/blob/644b9078b90a441a73da0ab9da66dc60d283033a/port/unix/omrsysinfo.c#L2670

In cgroup v1 (ref: https://www.kernel.org/doc/html/latest/scheduler/sched-bwc.html),
```
- File location: $mount_point/$subsytem_name[=cpu]/$cgroup_path/
- cpu.cfs_quota_us
- cpu.cfs_period_us
```

**... TRANSLATION from cgroup v1 to v2 ...**

In cgroup v2 (ref: https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#cgroup-v2-cpu),
```
- File location: $mount_point/$cgroup_path
- cpu.max format "$MAX $PERIOD". The $MAX value is equivalent to cpu.cfs_quota_us.
  If $MAX is "max", then it is equivalent to cpu.cfs_quota_us=-1, suggesting no bandwidth restriction.
- cpu.max format "$MAX $PERIOD". The $PERIOD value is equivalent to cpu.cfs_period_us.
```

**Comment by EricYangIBM** [comment]
If v2 mount point is not `/sys/fs/cgroup`, then 
```
If /sys/fs/cgroup/cgroup.controllers is present on your system, you are using v2, otherwise you are using v1.
```
is not valid. I noticed on my default fyre machine cgroup v1 and v2 hierarchies were present at the same time with the same processes on both (cgroup2 mounted at `/sys/fs/cgroup/unified`). 

For these hybrid systems, we can use the existing v1 logic. If v1 is not present, we can check for v2 by searching for `cgroup2` in `/proc/mounts` (which will also give us the mount point). So `isCgroupV2Available` will be

- search /proc/mounts for cgroup2, if found, save the mount point to a new global (like `PPG_cgroupSubsystemsAvailable`) and return TRUE

**Comment by babsingh** [comment]
### omrsysinfo_cgroup_subsystem_iterator_next

https://github.com/eclipse/omr/blob/644b9078b90a441a73da0ab9da66dc60d283033a/port/unix/omrsysinfo.c#L6225

**... TRANSLATION from cgroup v1 to v2 ...**

https://github.com/eclipse/omr/blob/644b9078b90a441a73da0ab9da66dc60d283033a/port/unix/omrsysinfo.c#L433

In cgroup v2,
```
ref cgroup v1 memory: https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt
ref cgroup v2 memory: https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#memory

static struct OMRCgroupSubsystemMetricMap omrCgroupMemoryMetricMap[] = {
	{ "memory.max", // cgroup v2 equivalent
	{ "memory.swap.max", // cgroup v2 equivalent
	{ "memory.current", // cgroup v2 equivalent
	{ "memory.swap.current", // cgroup v2 equivalent
	{ "memory.max_usage_in_bytes", // No cgroup v2 equivalent found
	{ "memory.memsw.max_usage_in_bytes", // No cgroup v2 equivalent found
	{ "memory.events" grep for "max %VALUE%" OR grep for "oom %VALUE%" // cgroup v2 equivalent
	{ "memory.memsw.failcnt", // No cgroup v2 equivalent found
	{ "memory.oom_control", // No cgroup v2 equivalent
};
```

Evaluation of `omrCgroupMemoryMetricMap` leads to **redundant operations** which are also performed in `retrieveLinuxCgroupMemoryStats`.


https://github.com/eclipse/omr/blob/644b9078b90a441a73da0ab9da66dc60d283033a/port/unix/omrsysinfo.c#L416

In cgroup v2,
```
ref cgroup v1 memory: https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt
ref cgroup v2 memory: https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#memory

static struct OMRCgroupMetricInfoElement oomControlMetricElementList[] = {
	"memory.oom_control -> "oom_kill_disable" // No cgroup v2 equivalent found but I think OOM killer is always enabled in cgroup v2
	"memory.oom_control -> "under_oom", NULL, FALSE } // No cgroup v2 equivalent found but this info can be derived from "memory.events" -> "oom_kill" or "oom_group_kill" stats
};
```

https://github.com/eclipse/omr/blob/644b9078b90a441a73da0ab9da66dc60d283033a/port/unix/omrsysinfo.c#L445

In cgroup v2,
```
ref cgroup v1 cpu: none found
ref cgroup v2 cpu: https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#cpu

static struct OMRCgroupSubsystemMetricMap omrCgroupCpuMetricMap[] = {
	{ cpu.max format "$MAX $PERIOD". The $PERIOD value is equivalent to cpu.cfs_period_us. // cgroup v2 equivalent
	{ cpu.max format "$MAX $PERIOD". The $MAX value is equivalent to cpu.cfs_quota_us. // cgroup v2 equivalent
	{ cpu.weight is equivalent to cpu.shares but it require some translation to derive CPU shares from cpu.weight // cgroup v2 equivalent
	{ cpu.stat // Stays the same for cgroup v2
};
```

Evaluation of `omrCgroupCpuMetricMap` leads to **redundant operations** which are also performed in `omrsysinfo_get_number_CPUs_by_type`.

https://github.com/eclipse/omr/blob/644b9078b90a441a73da0ab9da66dc60d283033a/port/unix/omrsysinfo.c#L421

In cgroup v2,
```
ref cgroup v1 cpu: none found
ref cgroup v2 cpu: https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#cpu

static struct OMRCgroupMetricInfoElement cpuStatMetricElementList[] = {
	{ "Period intervals elapsed count", "nr_periods" // Stays the same for cgroup v2
	{ "Throttled count", "nr_throttled", // Stays the same for cgroup v2
	{ "Total throttle time", "throttled_time", "nanoseconds", // Changed to "throttled_usec" in cgroup v2; we will need to convert it to nanoseconds
};
```

https://github.com/eclipse/omr/blob/644b9078b90a441a73da0ab9da66dc60d283033a/port/unix/omrsysinfo.c#L452

In cgroup v2,
```
ref cgroup v1 cpuset: https://www.kernel.org/doc/Documentation/cgroup-v1/cpusets.txt
ref cgroup v2 cpuset: https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#cpuset

static struct OMRCgroupSubsystemMetricMap omrCgroupCpusetMetricMap[] = {
	{ "cpuset.cpu_exclusive", // No cgroup v2 equivalent found
	{ "cpuset.mem_exclusive", // No cgroup v2 equivalent found
	{ "cpuset.cpus", // Stays the same for cgroup v2
	{ "cpuset.mems", // Stays the same for cgroup v2
};
```

**Comment by babsingh** [comment]
re https://github.com/eclipse/omr/issues/1281#issuecomment-1072702685:

You are correct that we need to dynamically derive the mount point.

https://www.kernel.org/doc/Documentation/cgroup-v1/cgroups.txt: based upon the mounting instructions, even cgroup v1 can have a different mount point. It becomes more complicated for cgroup v1 because each subsystem can have a unique mount point. so, even our existing implementation of cgroup v1 can malfunction if the default `/sys/fs/cgroup` mount point is overridden. Fixing cgroup v1 is not the priority.

**Option 1:** `/proc/mounts == /proc/self/mounts` since Linux 2.4.19
```
Format: https://man7.org/linux/man-pages/man5/fstab.5.html
cgroup /sys/fs/cgroup cgroup2 ro,nosuid,nodev,noexec,relatime 0 0
```

**Option 2:** `/proc/self/mounts`
```
Format: https://man7.org/linux/man-pages/man5/proc.5.html
device cgroup mounted on /sys/fs/cgroup with fstype cgroup2
```

**Option 3:** `/proc/self/mountinfo`
```
Format: https://man7.org/linux/man-pages/man5/proc.5.html
1223 1222 0:28 / /sys/fs/cgroup ro,nosuid,nodev,noexec,relatime - cgroup2 cgroup rw

(4)  root: the pathname of the directory in the filesystem which forms the root of this mount.
(5)  mount point: the pathname of the mount point relative to the process's root directory.
(9)  filesystem type: the filesystem type in the form "type[.subtype]".
(10) mount source: filesystem-specific information or "none".
(11) super options: per-superblock options (see [mount(2)](https://man7.org/linux/man-pages/man2/mount.2.html)).

(4)  0::$cgroup_path seen in /proc/self/cgroup
(5)  $mount_path to derive the full path 
(9)  cgroup or cgroup2
(10) cgroup
(11) for cgroup v1, it shows the subsystem name. eg: rw,memory
```

There is a only one cgroup2 file system so Options 1 and 2 should work fine for cgroup v2. But, they won't work with cgroup v1 because the subsystem details are missing.

**Option 3** should work with cgroup v1 and v2 since it has subsystem information at the end.

@EricYangIBM Can you verify if **Option 3** is the ideal approach?

**Comment by babsingh** [comment]
### Task List

Based upon all the previous comments, I have derived a list of tasks:
- [ ] **Task 1** -- Detect cgroup v1 vs. cgroup v2: https://github.com/eclipse/omr/issues/1281#issuecomment-1069231673; https://github.com/eclipse/omr/pull/6422.
- [ ] **Task 2** -- Update `omrsysinfo_cgroup_is_system_available` for cgroup v2: https://github.com/eclipse/omr/issues/1281#issuecomment-1072527260.
- [ ] **Task 3** -- Update `getAbsolutePathOfCgroupSubsystemFile` to locate cgroup v2 related files. For cgroup v1, the file path is `$mount_point/$subsystem_name[/$cgroup_path/$file_name`. For cgroup v2, the file path will be `$mount_point/$cgroup_path/$file_name`. https://github.com/eclipse/omr/issues/1281#issuecomment-1072609979.
- [ ] **Task 4** -- Update `retrieveLinuxCgroupMemoryStats` for cgroup v2: https://github.com/eclipse/omr/issues/1281#issuecomment-1072609979.
- [ ] **Task 5** -- Update `omrsysinfo_get_number_CPUs_by_type` for cgroup v2: https://github.com/eclipse/omr/issues/1281#issuecomment-1072637216.
- [ ] **Task 6** -- Dynamically derive `MOUNT_POINT` for cgroup v1 and cgroup v2: https://github.com/eclipse/omr/issues/1281#issuecomment-1072782528.
- [ ] **Task 7** -- Avoid redundant code https://github.com/eclipse/omr/issues/1281#issuecomment-1072740460:
    - Evaluation of `omrCgroupMemoryMetricMap` leads to **redundant operations** which are also performed in `retrieveLinuxCgroupMemoryStats`.
    - Evaluation of `omrCgroupCpuMetricMap` leads to **redundant operations** which are also performed in `omrsysinfo_get_number_CPUs_by_type`.
- [ ] **Task 8** -- Update `omrsysinfo_cgroup_subsystem_iterator_next` for cgroup v2: https://github.com/eclipse/omr/issues/1281#issuecomment-1072740460.
- [ ] **Task 9** (optional) -- Use `J9HashTable` instead of a list. It will give better performance. 

Until **Task 6**, let the mount point be static: `define OMR_CGROUP_MOUNT_POINT "/sys/fs/cgroup"`. I am assuming that most cgroups will adopt this default mount point. I think prioritizing the tasks in the above order is the most efficient way of adding cgroup v2 support and fixing the existing cgroup v1 support.

**Comment by babsingh** [comment]
> Have we had any issues with assuming the v1 mount point is /sys/fs/cgroup? 

In this case, the existence of issues should not be a deciding factor on whether we should fix the cgroup v1 code.

a. First, if this API malfunctions, it can go unnoticed. Symptoms are generally non-blocking. The app may continue to function without a noticeable failure such as incorrect settings or a perf regression. 

b. Second, there is insufficient testing for the cgroup API. I only noticed one test: `sysinfo_cgroup_get_memlimit`. The testing lacks in functional coverage. Also, we don't vary the environment, which is an important element in testing this API: 1) Linux kernel host with cgroup v1, 2) Linux kernel host with cgroup v2, 3) Docker envionment with cgroup v1, and 4) Docker environment with cgroup v2.

**Comment by babsingh** [comment]
> Based on the kernel documentation and man page for v1, I get the impression that mounting at /sys/fs/cgroup is a given

The above statement is incorrect. The below commands allow the mount point to be changed for cgroup v1:
```
umount /sys/fs/cgroup/pids
mkdir /tmp/sys/fs/cgroup/pids
mount -t cgroup -o pids none /tmp/sys/fs/cgroup/pids

cat /proc/960/mountinfo # Uses the new mount point

46 32 0:40 / /tmp/sys/fs/cgroup/pids rw,relatime shared:20 - cgroup none rw,pids
```

Based upon the above output, we will probably need to fix the cgroup v1 code. 

@DanHeidinga @ashu-mehra In the current implementation, why a static mount point (`/sys/fs/cgroup`) is assumed for the cgroup API? Are we expected to support non-default mount points?

**Comment by ashu-mehra** [comment]
@babsingh yes, the mount point can be changed, but in my experience I have not seen it mounted at any different location. It is more or less a standard practice to mount it at `/sys/fs/cgroup`, at least for systemd based distros.

One of the recommendation in https://www.freedesktop.org/wiki/Software/systemd/PaxControlGroups/ says:

> The cgroup mount point for a specific hierarchy is /sys/fs/cgroup/$CONTROLLER/. (Example: /sys/fs/cgroup/cpu for the "cpu" controller). In your application you are welcome to rely on these standardized mount points, and it is not necessary to dynamically determine the current mount point via /proc/self/mountinfo (but if you do, that's of course fine, too). Note that /sys/fs/cgroup/$CONTROLLER/ might actually just be a symlink to some other mount point (see below).

Something similar is mentioned in this [blog post](http://0pointer.de/blog/projects/cgroups-vs-cgroups.html):

> On boot, systemd will mount all controller hierarchies it finds enabled in the kernel to individual directories below /sys/fs/cgroup/. This is the official place where kernel controllers are mounted to these days. The /sys/fs/cgroup/ mount point in the kernel was created precisely for this purpose.

Hope this helps.
